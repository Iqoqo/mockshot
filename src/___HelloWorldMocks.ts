export class HelloWorldMocks {
    static bar;

    static bar(mock: "no-valid-params" | "success"): any {
        switch (mock) {
            case "no-valid-params":
                return {
                  "error": "yes"
                }
            case "success":
                return {
                  "api": {
                    "data": "123"
                  }
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }

    static shape;

    static shape(mock: "success"): any {
        switch (mock) {
            case "success":
                return {
                  "data": {
                    "_id": "1234"
                  },
                  "id": "1234sddd56",
                  "some": "value"
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }

    static foo;

    static foo(mock: "success"): any {
        switch (mock) {
            case "success":
                return {
                  "foo": "barr"
                }
            default:
                throw Error("Unknown mock: "+mock);
        }
    }
}
