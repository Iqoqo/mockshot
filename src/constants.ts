export const MockshotTag = "[mockshot]";
export const Success = "success";
