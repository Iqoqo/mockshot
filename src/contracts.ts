export interface ISnapshot {
  key: string;
  packageName?: string;
  filePath: string;
  data: any;
}

export interface IApiSnapshot extends ISnapshot {
  data: IApiSnapData;
}

export interface IClassSnapshot extends ISnapshot {
  data: IClassSnapData;
}

export interface IApiSnapData extends IApiSnapDataBase {
  mockName: string;
}

export interface IClassSnapData {
  className: string;
  methodName: string;
  mockName: string;
  mock: any;
}

export interface IApiSnapDataBase {
  url: string;
  httpMethod: "post" | "get" | "put" | "delete" | "patch";
  mock: {
    statusCode: number;
    body: any;
    error?: string;
  };
}

export type MatcherReturn = {
  pass: boolean;
  message(): string | (() => string);
};
