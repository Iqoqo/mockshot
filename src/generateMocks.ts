import {glob} from "glob";
import * as path from "path";
import Project from "ts-simple-ast";
import fs from "fs";
import { SnapshotState } from "jest-snapshot";
import pretty from 'json-pretty'

export async function generateMocks() {
  const project = new Project({
    compilerOptions: { outDir: "dist/mocks", declaration: true }
  });

  glob("**/*.snap", async function(er, files) {
    // files is an array of filenames.
    // If the `nonull` option is set, and nothing
    // was found, then files is ["**/*.js"]
    // er is an error object or null.
    const mockDef = {} as any;

    files.forEach(file => {
      const absolutePath = path.resolve(file);
      const state = new SnapshotState(absolutePath, {
        snapshotPath: absolutePath
      });
      const data = state._snapshotData;
      const keys = Object.keys(data);

      keys.forEach(async key => {
        if (key.indexOf("[mockshot]") < 0) {
          return;
        }
        const snapshot = JSON.parse(state._snapshotData[key]);
        let classDef = mockDef[snapshot.className];
        if (!classDef) {
          classDef = {};
          mockDef[snapshot.className] = classDef;
        }
        let methodDef = classDef[snapshot.methodName];
        if (!methodDef) {
          methodDef = {};
          classDef[snapshot.methodName] = methodDef;
        }

        let mockName = methodDef[snapshot.mockName];
        if (!mockName) {
          methodDef[snapshot.mockName] = snapshot.mock;
        } else {
          throw Error(
            "Duplicate mock name for method " +
              snapshot.methodName +
              ": " +
              snapshot.mockName
          );
        }
      });
    });

    const classNames = Object.keys(mockDef);
    classNames.forEach(async className => {
      const mockClassName = className+'Mocks';

      console.log('Creating class', mockClassName);

      const mockFileName = `mocks/${mockClassName}.ts`;

      let myClassFile;
      try {
        await fs.unlinkSync(mockFileName);
        console.log("File exists", mockFileName, "removing...");
      } catch (ex) {
      }

      myClassFile = project.createSourceFile(mockFileName);

      const classDeclaration = myClassFile.addClass({
        name: mockClassName
      });

      classDeclaration.setIsExported(true);

      const methods = mockDef[className];

      const methodNames = Object.keys(methods);

      methodNames.forEach(methodName => {
        console.log('Creating method mocks', methodName);

        const mocks = methods[methodName];
        const mockNames = Object.keys(mocks);
        let mockTypes = mockNames.map(value => {
          return `"${value}"`;
        });
        const types = mockTypes.join(" | ");

        const method = classDeclaration.addMethod({
          isStatic: true,
          parameters: [{ name: "mock", type: types }],
          name: methodName,
          returnType: "any"
        });

        method.setBodyText(writer =>
          writer.write("switch (mock)").block(() => {
            mockNames.forEach(mockName => {
              writer.write(`case "${mockName}":`).indentBlock(()=>{
                writer.write(`return ${pretty(mocks[mockName])}`);
              })
            });
            writer.write(`default:`).indentBlock(()=>{
              writer.write(`throw Error("Unknown mock: "+mock);`);
            })
          })
        );
      });
    });

    await project.save();

    console.log('Saved Typescript source...emitting vanilla js');
    await project.emit();
    console.log('Done!');
  });
}
