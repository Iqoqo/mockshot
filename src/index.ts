export * from "./matchers";
export * from "./generators";
export { run } from "./run";
