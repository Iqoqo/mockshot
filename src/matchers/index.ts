import stringify from 'json-stable-stringify';
import { toMatchApiMock } from "./ApiMockMatcher";
import { toMatchClassMock } from "./ClassMockMatcher";

declare global {
    namespace jest {
        // tslint:disable-next-line:interface-name
        interface Matchers<R> {
            toMatchClassMock(
                className: string,
                methodName: string,
                mockName: string,
                ignoredKeyPaths?: string[]
            ): R;
            toMatchApiMock(
                response: object,
                mockName: string,
            ): R;
        }
    }
}

expect.addSnapshotSerializer({
    test: val => val.mock,
    print: val => stringify(val, { space: '  ' })
});

const toMatchMock = toMatchClassMock

expect.extend({ toMatchMock, toMatchClassMock, toMatchApiMock });

export { toMatchMock, toMatchClassMock, toMatchApiMock };
