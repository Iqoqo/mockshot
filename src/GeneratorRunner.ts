import fs from "fs";
import path from "path";
import Project, { SourceFile } from "ts-simple-ast";
import { ISnapshot } from "./contracts";
import { MockGenerator } from "./generators/base";

export class GeneratorRunner {
  private project: Project;

  constructor(
    private outDir: string,
    private generators: MockGenerator[],
    private snapshots: ISnapshot[]
  ) {
    this.project = new Project();
  }

  async run() {
    this.generators.forEach(this.runGenerator);
    this.project.saveSync();
    await this.project.emit();
  }

  private runGenerator = (generator: MockGenerator) => {
    generator.generate(this.createSourceFile, this.snapshots);
  };

  private createSourceFile = (mockFileName: string): SourceFile => {
    fs.unlinkSync(path.join(this.outDir, mockFileName));
    console.log("File exists", mockFileName, "removing...");
    return this.project.createSourceFile(path.join(this.outDir, mockFileName));
  };
}
