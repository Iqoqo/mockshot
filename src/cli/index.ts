import { ApiGenerator } from "../generators/ApiGenerator";
import { GeneratorRunner } from "../GeneratorRunner";
import { SnapshotFetcher } from "./util";

export async function run(args: string[]) {
  const outputDir = args[0];

  const snapshots = SnapshotFetcher.getSnapshots();

  const generators = [new ApiGenerator()];

  const generatorRunner = new GeneratorRunner(outputDir, generators, snapshots);
  await generatorRunner.run();
}
