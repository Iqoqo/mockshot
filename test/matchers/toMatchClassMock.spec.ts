import "../../src/matchers";

class HelloWorld {
  bar(mock: "success"): any {
    switch (mock) {
      case "success":
        return {
          "data": {
            "_id": "1234"
          },
          "id": "1234sddd56",
          "some": "value"
        }
      default:
        throw Error("Unknown mock: " + mock);
    }
  }

  foo(mock: "success"): any {
    switch (mock) {
      case "success":
        return {
          "foo": "barr"
        }
      default:
        return "unknown";
    }
  }

  complexArray(mock: "complexArray1" | "complexArray2" | "complexArray3"): any {
    switch (mock) {
      case "complexArray1": {
        return {
          "data": [{ "id": 999 }, { "id": 567 }]
        }
      }
      case "complexArray2": {
        return {
          "data": {
            "level1": [{ "id": 123 }, { "id": 456 }]
          }
        }
      }
      case "complexArray3": {
        return {
          "data": [{
            "level1": [{ "id": 123 }, { "id": 456 }]
          }, {
            "level1": [{ "id": 789 }]
          }]
        }
      }
    }
  }
}

const hello = new HelloWorld();

describe("toMatchMock", () => {
  it("Should return correct schema", () => {
    expect(hello.foo("success")).toMatchMock(HelloWorld.name, "foo", "success");
  });

  it("Should match a default parameter", () => {
    expect(hello.foo("default-param")).toMatchMock(HelloWorld.name, "foo", "default-param");
  });

  it("Should match mock with ignoredKeyPaths", () => {
    expect(hello.bar("success")).toMatchMock(HelloWorld.name, "bar", "success", ["id", "data._id"]);
  });

  it.only("Should igone fields within an array", () => {
    expect(hello.complexArray("complexArray1")).toMatchMock(HelloWorld.name, "complexArray", "complexArray1", ["data[id]"]);
    //NOT SUPPORTED YET
    // expect(hello.complexArray("complexArray2")).toMatchMock(HelloWorld.name, "complexArray", "complexArray2", ["data.level1[id]"]);
    // expect(hello.complexArray("complexArray3")).toMatchMock(HelloWorld.name, "complexArray", "complexArray3", ["data[level1[id]]"]);
  })

});
