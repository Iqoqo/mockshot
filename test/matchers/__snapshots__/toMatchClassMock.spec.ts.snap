// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchMock Should igone fields within an array: [mockshot] [ClassSnap] [[HelloWorld complexArray complexArray1]] 1`] = `
{
  "className": "HelloWorld",
  "methodName": "complexArray",
  "mock": {
    "data": [
      {
        "id": 123
      },
      {
        "id": 567
      }
    ]
  },
  "mockName": "complexArray1"
}
`;

exports[`toMatchMock Should igone fields within an array: [mockshot] [ClassSnap] [[HelloWorld complexArray complexArray2]] 1`] = `
{
  "className": "HelloWorld",
  "methodName": "complexArray",
  "mock": {
    "data": {
      "level1": [
        {
          "id": 123
        },
        {
          "id": 456
        }
      ]
    }
  },
  "mockName": "complexArray2"
}
`;

exports[`toMatchMock Should igone fields within an array: [mockshot] [ClassSnap] [[HelloWorld complexArray complexArray3]] 1`] = `
{
  "className": "HelloWorld",
  "methodName": "complexArray",
  "mock": {
    "data": [
      {
        "level1": [
          {
            "id": 123
          },
          {
            "id": 456
          }
        ]
      },
      {
        "level1": [
          {
            "id": 789
          }
        ]
      }
    ]
  },
  "mockName": "complexArray3"
}
`;

exports[`toMatchMock Should match a default parameter: [mockshot] [ClassSnap] [[HelloWorld foo default-param]] 1`] = `
{
  "className": "HelloWorld",
  "methodName": "foo",
  "mock": "unknown",
  "mockName": "default-param"
}
`;

exports[`toMatchMock Should match mock with ignoredKeyPaths: [mockshot] [ClassSnap] [[HelloWorld bar success]] 1`] = `
{
  "className": "HelloWorld",
  "methodName": "bar",
  "mock": {
    "data": {
      "_id": "1234"
    },
    "id": "1234sddd56",
    "some": "value"
  },
  "mockName": "success"
}
`;

exports[`toMatchMock Should return correct schema: [mockshot] [ClassSnap] [[HelloWorld foo success]] 1`] = `
{
  "className": "HelloWorld",
  "methodName": "foo",
  "mock": {
    "foo": "barr"
  },
  "mockName": "success"
}
`;
